onnxruntime/core/framework/utils.h

```diff
@@ -87,69 +87,254 @@ void DumpNodeInputs(const OpKernelContext& context, const Node& node);
+    case ONNX_NAMESPACE::TensorProto_DataType_UINT8:                       \
+      retval = function<uint8_t>(__VA_ARGS__);                             \
+      break;                                                               \
+    case ONNX_NAMESPACE::TensorProto_DataType_UINT16:                      \
+      retval = function<uint16_t>(__VA_ARGS__);                            \
+      break;                                                               \
+    case ONNX_NAMESPACE::TensorProto_DataType_INT16:                       \
+      retval = function<int16_t>(__VA_ARGS__);                             \
+      break;                                                               \
+    case ONNX_NAMESPACE::TensorProto_DataType_INT32:                       \
+      retval = function<int32_t>(__VA_ARGS__);                             \
+      break;                                                               \
+    case ONNX_NAMESPACE::TensorProto_DataType_UINT32:                      \
+      retval = function<uint32_t>(__VA_ARGS__);                            \
+      break;                                                               \
+    case ONNX_NAMESPACE::TensorProto_DataType_INT64:                       \
+      retval = function<int64_t>(__VA_ARGS__);                             \
+      break;                                                               \
+    case ONNX_NAMESPACE::TensorProto_DataType_UINT64:                      \
+      retval = function<uint64_t>(__VA_ARGS__);                            \
+      break;                                                               \
+    case ONNX_NAMESPACE::TensorProto_DataType_FLOAT16:                     \
+      retval = function<MLFloat16>(__VA_ARGS__);                           \
+      break;                                                               \
+    case ONNX_NAMESPACE::TensorProto_DataType_BFLOAT16:                    \
+      retval = function<BFloat16>(__VA_ARGS__);                            \
+      break;                                                               \
+    default:                                                               \
+      ORT_ENFORCE(false, "Unknown tensor type of ", tensor_type);          \
+  }
+
+inline bool IsDataTypeString(MLDataType dt_type) {
+  auto prim_type = dt_type->AsPrimitiveDataType();
+  return (prim_type != nullptr && prim_type->GetTensorElementType() == ONNX_NAMESPACE::TensorProto_DataType_STRING);
+}
+
+template<class T>
+inline bool IsPrimDataType(MLDataType dt_type) {
+  auto prim_type = dt_type->AsPrimitiveDataType();
+  return (prim_type != nullptr && prim_type->GetTensorElementType() == ToTensorDataType<T>());
+}
+
+// Use after AsPrimitiveDataType() is successful
+template<class T>
+inline bool IsPrimDataType (const PrimitiveDataTypeBase* prim_type) {
+  assert(prim_type != nullptr);
+  return prim_type->GetTensorElementType() == ToTensorDataType<T>();
```

Comment:
Can prim_type be a reference instead of having this assert?

Reply:
In theory it could be a ref, but in practice it would be counterproductive. 1) It would be inconvenient bc MLDataType is a ptr everywhere and the user would be forced to convert it to a reference 2) The fact it would be a reference would not prevent it to be a nullptr, but we would loose the ability to assert on it.

---
In reply to: [341918377](https://github.com/microsoft/onnxruntime/pull/2307#discussion_r341918377) [](ancestors = 341918377)